### DISCUSSION TITLE
How to handle Input/Output properties when linked

### DISCUSSION THREAD

**Discussion by dibli-goost** [bnpr/Malt#414] [Q&A] [answered]
I feel like the best use case with Render Trees is to create a big project containing a bunch of Render Trees you can use in different scenarios. So for example you have a Character Render Tree for use in Character projects, with all the screen passes you want for that scenario, you may not need Screen Passes for VFX and stuff as it might just get in the way of performance. Meanwhile, for VFX projects you might want a VFX Render Tree, with various VFX screen passes, sans stuff you want for characters.. and so on.

Linking entire render trees of course means that you can make changes in one version, and it apply everywhere else. Very useful stuff for large blend file counts. Most of this is working fine, I've been doing it for awhile for Line Art. But the big issue I'm finding is with AOVs.

I might add a Bloom Color and Bloom Radius output to the Main Pass Output, but that isn't linked *with* a Render Tree. So when you link it to a new project, you have to manually add those outputs to the Main Pass in that new one.

Even when linking entire Mesh Node Trees from one project to others, the Main Pass Outputs aren't preserved. Here I have a Bloom Color output in this project.
![image](https://user-images.githubusercontent.com/31767887/193482998-e49924c3-20d8-4c68-84bb-57a7adedcaef.png)
But it's not there when linked.
![image](https://user-images.githubusercontent.com/31767887/193483057-6466c2bc-eb7d-48f4-8ba3-a2d946a979db.png)

I don't think this is a bug, is it? It seems like a lack of a feature to me.

Don't be fooled by when these properties are connected to noodles btw. While it looks like the property exists. It's not actually listed in the N panel
![image](https://user-images.githubusercontent.com/31767887/193483168-0dcd793c-666e-49ef-94c7-92a0ee434536.png)

-----

I'm not really sure what the best way around this is. The first suggestion that came to mind would be to orient the Main Pass Outputs around the Main Pass node in the Render Layers Tree. If an output exists there: consider the property real in all node trees? This might be difficult to implement though and could cause issues in the future.

Another suggestion, although it wouldn't fix the issue of needing to manually add the outputs for Render Trees, would be for when linking general Node Trees from other blends. Maybe Malt could detect what outputs are being used on the Main Pass there and automatically add them to the Main Pass in the new blend.

**Comment by pragma37** [comment, marked as answer]
There are two kinds of custom IO:
- Per Tree. Used in Screen shaders. These ones store the custom IO "signature" in the node tree itself, so they should work fine when linking.
- Per Pass Type. Used in Mesh shaders. Here the custom IO must be shared by all the shaders in the scene, **so the custom IO "signature" is stored in the World**. 
This means that if, instead of linking the Render nodes, you link the World instead (which contains the Render nodes anyway), custom IO will be linked as well.

I just made the test and I had to click the *pipeline reload* button before it worked, so don't be surprised if you find other bugs, but in theory it should work.

**Reply by pragma37** [reply]
And sorry for the late reply!
It has been a very busy week for me and this flew under my radar.

**Reply by dibli-goost** [reply]
Ah I see! That does work... didn't think to link the world itself.

Only problem that I have with this functionality so far is that you can't control the viewport resolution if you link the world. I find it to be pretty beneficial overall.. some way to override the link there would be nice. It's not really that big of an issue though.
